Use Closure and type definition environment during typechecking to check unbound types

litchipi opened the pull request:
Add a `typedef` field to the typechecker environment, that will check if a variable  is defined in the environment before raising an error.
Also removes anything related to types in the `parser`

Useful for future fix of PR #464 
Can be also used for future modifications related to #202

litchipi commented:
As this PR doesn't implement usage of such type environment during the typechecking, **this is just a base draft** that is currently modified and adapted while developing user-defined types on top of it.

yannham commented:
> As this PR doesn't implement usage of such type environment during the typechecking, this is just a base draft that is currently modified and adapted while developing user-defined types on top of it.

In this case, you can set it as a Draft PR, to indicate that it is not ready for review yet. Below the reviewers on the right panel you have a "convert to draft" button.

litchipi commented:
Implemented Closures, should work with user-defined types (#464) in typechecking.
Currently raises a stackoverflow error on `cargo test` for `eq.ncl` test, WIP fix

litchipi commented:
As it was a long rabbit hole, and a lot of changes occured since then, I had to clean & rebase the working implementation into a single commit on top of master.
That may still have hacky things and things no longer useful.

I will apply fixup commits on top of the branch and squash up everything into a single one before merging once the review is OK

litchipi commented:
With the PR #500, the CI should work.
I just squashed all the commits and fixups into 2 clean commits:
- To implement the Closure in typechecks
- To modify the `test/pass/eq.ncl` to prevent it from overflowing the stack.

Note that the latter will need a proper fix later, by rewriting some parts of the code.

litchipi commented:
Removed the quick fix for the failing test, as it has its own issue (#502) that requires to be at least quickly solved before this PR can be merged (if not, a lot of future changes will make the tests fail).

yannham commented on src/typecheck/mod.rs:
```diff
@@ -1326,121 +1659,189 @@ pub fn unify_(
+        absty
+    } else {
+        unreachable!("Not concreat abstype unify_rows (2)")
+    };
```
Rather than having this, you can just pass the original abs type and the environment separated. Not all arguments need to be passed as packed closures. E.g:

```
unify_rows(
  state: &mut State,
  t1: AbsType<Box<TypeWrapper>>,
  env1: Environment,
  t2: AbsType...
```

Then you don't have those nasty `unreachable` anymore.

yannham commented on src/typecheck/mod.rs:
```diff
@@ -1452,30 +1853,49 @@ pub fn unify_rows(
+                ty.clone(),
+                Closure {
+                    ty: TypeWrapper::Concrete(r2),
+                    env: env1.clone(),
```
while all those clones? Can't we reuse the environment `env1` directly?

litchipi replied:
No because `env1` is used later, so using it without a `clone` would move it to a new owner.
It may appear again some unnecessary clones through the code that are over-protections for that kind of problems, but I don't think this is one of them

yannham commented on src/typecheck/mod.rs:
```diff
@@ -1856,175 +2306,191 @@ pub fn get_uop_type(
 pub fn get_bop_type(
     state: &mut State,
     op: &BinaryOp,
-) -> Result<(TypeWrapper, TypeWrapper, TypeWrapper), TypecheckError> {
```
Types returned by this function shouldn't depend on the environment. So it should return `TypeWrapper`s as before, and the caller would map `Closure::atomic` over the triple. If it's simpler it can return `Closure` by doing it itself, but it shouldn't take an environment as a parameter, and put it into the closure.

yannham reviewed (commented):
It's not a full review, but a first series of comments

yannham commented on src/typecheck/mod.rs:
```diff
@@ -474,19 +592,31 @@ where
             false,
             t,
-            ty.clone(),
+            Closure {
```
Sorry, this comment is at the wrong place it seems. I meant instead of doing
```
Closure {
  ty: ty.clone(),
  env: Env::new()
}
```

litchipi commented on src/typecheck/mod.rs:
```diff
@@ -339,65 +385,75 @@ impl UnifError {
+        let mut env = TypeEnvironment::new();
+        for (id, tw) in g.iter() {
+            env.insert(id.clone(), tw.into());
+        }
```
I don't understand what you mean, the `&GlobalEnvironment` argument is still passed in the `type_check` (and other) functions for this purpose

litchipi replied:
Note that right now we are not dealing with type aliases yet, even if this PR is made for this feature request.

I don't understand the difference between the local typing environment and the local typealias environment, and why it needs to be 2 separated environments when we could merge the two of them and use them inside Closures.

The reason I didn't merged them is that the local "typealias" environment is made for explicitly defined type bindings (and to add a simple layer instead of breaking the existing system)

litchipi commented on lsp/nls/src/linearization.rs:
```diff
@@ -380,7 +380,12 @@ impl Linearizer<BuildingResource, (UnifTable, HashMap<usize, Ident>)> for Analys
+                        &table,
+                        &reported_names,
+                        &mut NameReg::new(),
+                        Closure::new(ty, Envs::new()),
```
Every `Closure`-related stuff had been cleaned from linearization in the last version, let's state on what is needed there as it's totally blurred for me right now.

litchipi commented:
I added a bit more explanations on the commit in the commit message (as this is a somewhat big one).
3 points have to be addressed now:
- There's a bunch of renames of variables (that makes more sense considering their type), that generate a bit of useless diff. Should they be removed or not ?
- Nothing significant has been changed in the `lsp` for now, and maybe it should use Closures and typing environment as well. Would be cool if @ysndr can look at it and see how it could be used (or not) for lsp
- I modified the (draft I guess ?) of `LetPattern` in `typechecking/mod.rs` as I think it should be using Closures, but it may be wrong (and I guess this feature is not finished yet), a feedback from @francois-caddet could be cool on it

I tried to reduce to the minimum the propagation of `Envs`, only where it's useful, and simplify anything that was possibly simplified using them. A general review on the code and it's integration in the software architecture is needed now

litchipi commented on src/typecheck/mod.rs:
```diff
@@ -450,6 +469,85 @@ pub struct State<'a> {
+
+    pub fn flatten_type(self) -> TypeWrapper {
+        let clos = self.clone().resolve_typealias().unwrap_or(self);
+        match clos.ty {
```
Artefact from previous devs, cleaned that
